internal/appsec/waf_test.go

```diff
@@ -495,6 +502,149 @@ func TestAPISecurity(t *testing.T) {
 	})
 }
 
+func prepareSQLDB(nbEntries int) (*sql.DB, error) {
+	const tables = `
+CREATE TABLE user (
+   id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
+   name  text NOT NULL,
+   email text NOT NULL,
+   password text NOT NULL
+);
+CREATE TABLE product (
+   id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
+   name  text NOT NULL,
+   category  text NOT NULL,
+   price  int NOT NULL
+);
+`
+	db, err := sqltrace.Open("sqlite3", ":memory:")
+	if err != nil {
+		log.Fatalln("unexpected sql.Open error:", err)
+	}
+	sqltrace.Register("sqlite3", &sqlite3.SQLiteDriver{})
+
+	if _, err := db.Exec(tables); err != nil {
+		return nil, err
+	}
+
+	for i := 0; i < nbEntries; i++ {
+		_, err := db.Exec(
+			"INSERT INTO user (name, email, password) VALUES (?, ?, ?)",
+			fmt.Sprintf("User#%d", i),
+			fmt.Sprintf("user%d@mail.com", i),
+			fmt.Sprintf("secret-password#%d", i))
+		if err != nil {
+			return nil, err
+		}
+
+		_, err = db.Exec(
+			"INSERT INTO product (name, category, price) VALUES (?, ?, ?)",
+			fmt.Sprintf("Product %d", i),
+			"sneaker",
+			rand.Intn(500))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return db, nil
+}
+
+func TestRASPSQLi(t *testing.T) {
+	t.Setenv("DD_APPSEC_RULES", "testdata/rasp.json")
+	appsec.Start()
+	defer appsec.Stop()
+
+	if !appsec.RASPEnabled() {
+		t.Skip("RASP needs to be enabled for this test")
+	}
+	db, err := prepareSQLDB(10)
+	require.NoError(t, err)
+
+	// Setup the http server
+	mux := httptrace.NewServeMux()
+	mux.HandleFunc("/query", func(w http.ResponseWriter, r *http.Request) {
+		// Subsequent spans inherit their parent from context.
+		q := r.URL.Query().Get("query")
+		rows, err := db.QueryContext(r.Context(), q)
+		if events.IsSecurityError(err) {
+			w.WriteHeader(403)
+			w.Write([]byte(err.Error()))
+			return
+		}
+		if err == nil {
+			rows.Close()
+		}
+		w.Write([]byte("Hello World!\n"))
+	})
+	mux.HandleFunc("/exec", func(w http.ResponseWriter, r *http.Request) {
+		// Subsequent spans inherit their parent from context.
+		q := r.URL.Query().Get("query")
+		_, err := db.ExecContext(r.Context(), q)
+		if events.IsSecurityError(err) {
+			w.WriteHeader(403)
+			w.Write([]byte(err.Error()))
+			return
+		}
+		w.Write([]byte("Hello World!\n"))
+	})
+	srv := httptest.NewServer(mux)
+	defer srv.Close()
+
+	for name, tc := range map[string]struct {
+		query string
+		err   error
+	}{
+		"no-error": {
+			query: "SELECT%201",
+		},
+		"injection/SELECT": {
+			query: "SELECT%20%2A%20FROM%20users%20WHERE%20user%3D%22%22%20UNION%20ALL%20SELECT%20NULL%2Cversion%28%29--",
+			err:   &events.BlockingSecurityEvent{},
+		},
+		"injection/UPDATE": {
+			query: "UPDATE%20users%20SET%20pwd%20%3D%20%22root%22%20WHERE%20id%20%3D%20%22%22%20OR%201%20%3D%201--",
+			err:   &events.BlockingSecurityEvent{},
+		},
+		"injection/EXEC": {
+			query: "EXEC%20version%28%29%3B%20DROP%20TABLE%20users--",
+			err:   &events.BlockingSecurityEvent{},
```

Comment:
Can you use [`url.QueryEscape`](https://pkg.go.dev/net/url?#QueryEscape) to make the tests actually readable ?